sel4-ftrace-merge/ftrace-merge.c

```diff
@@ -0,0 +1,863 @@
+
+struct flyrecord_wrap read_flyrecords(FILE *f, const uint32_t n_cpu, size_t *_size)
+{
+	int err = 0;
+	const size_t size = sizeof(struct flyrecord) * n_cpu;
+	struct flyrecord *flyrecords;
+
+	print_dbg("\n%s: call\n", __func__);
+
+	flyrecords = malloc(size);
+	assert(flyrecords);
+
+	err = fread(flyrecords, 1, size, f);
+	assert(err == size);
+
+	for (int i = 0; i != n_cpu; ++i)
+		print_dbg("CPU %02d -- offset: 0x%lx size: 0x%lx\n", i, flyrecords[i].offset, flyrecords[i].size);
+
+	struct flyrecord_wrap wrap = {
+		.size = size,
+		.flyrecords = flyrecords
+	};
+	*_size += wrap.size;
+	assert(*_size == ftell(f));
+	return wrap;
+}
+
+struct padding_wrap read_padding(FILE *f, struct ftrace_file *ftrace)
+{
+	int err;
+	char *data;
+	size_t size;
+	const struct flyrecord *flyrecords = ftrace->flyrecords.flyrecords;
+
+	print_dbg("\n%s: call\n", __func__);
+
+	/* #TODO: assumed that we have at least one CPU */
+	size = flyrecords[0].offset - ftell(f);
+	print_dbg("padding size: 0x%lx\n", size);
+
+	data = malloc(size);
+	assert(data);
+
+	err = fread(data, 1, size, f);
+	assert(err == size);
+
+	struct padding_wrap wrap = {
+		.size = size,
+		.data = data
+	};
+	ftrace->size += wrap.size;
+	assert(ftrace->size == ftell(f));
+	return wrap;
+}
+
+void check_data_integrity(FILE *f, const struct ftrace_file *const ftrace)
+{
+	size_t size;
+
+	const long position = ftell(f);
+
+	fseek(f, 0, SEEK_END);
+
+	const long end_position = ftell(f);
+
+	fseek(f, position, SEEK_SET);
+
+	print_dbg("\n%s: call\n", __func__);
+
+	size = ftrace->event_formats.size +
+		ftrace->flyrecords.size +
+		ftrace->ftrace_event_formats.size +
+		ftrace->header_end.size +
+		ftrace->header_event_format.size +
+		ftrace->header_page_format.size +
+		ftrace->initial.size +
+		ftrace->kallsyms_info.size +
+		ftrace->printk_info.size +
+		ftrace->process_info.size +
+		ftrace->padding.size +
+		ftrace->rest_data.size;
+
+	print_dbg("FTRACE size: %lu expected size: %lu\n", size, end_position);
+	assert(end_position == size);
+	print_dbg("FTRACE size: %lu expected size: %lu\n", size, ftrace->size);
+	assert(size == ftrace->size);
+}
+
+struct rest_data_wrap read_rest_data(FILE *f, struct ftrace_file *ftrace)
+{
+	int err;
+	char *data;
+	size_t size, expected_size;
+	const uint32_t n_cpu = ftrace->header_end.header_end->n_cpu;
+	const struct flyrecord *flyrecords = ftrace->flyrecords.flyrecords;
+
+	print_dbg("\n%s: call\n", __func__);
+
+	const long position = ftell(f);
+
+	fseek(f, 0, SEEK_END);
+
+	const long end_position = ftell(f);
+
+	fseek(f, position, SEEK_SET);
+
+	size = end_position - position;
+	expected_size = flyrecords[n_cpu - 1].offset + flyrecords[n_cpu - 1].size - flyrecords[0].offset;
+	print_dbg("size of Rest: %lu expected size: %lu\n", size, expected_size);
+	assert(size == expected_size);
+
+	data = malloc(size);
+	assert(data);
+
+	err = fread(data, 1, size, f);
+	assert(err == size);
+
+	struct rest_data_wrap wrap = {
+		.size = size,
+		.data = data
```

Comment:
Add comma even if it's the last member, this way if you introduce new members to structs, you can define them here and still the patch would not need to include the addition of comma here.

Reply:
Fixed.